### DISCUSSION TITLE
Is there build support in Linux?

### DISCUSSION THREAD

**Discussion by derrikF** [ScanTailor-Advanced/scantailor-advanced#24] [Q&A] [answered]
I'm trying to build in Linux Mint 21 (Ubuntu 22.04). I tried in Linux Mint 20.3 and Manjaro

Compilation fails everywhere. Tried both with Qt5.15 and Qt6

I wanted to ask - is there only support for Windows builds here?

It doesn't seem to be able to build in Linux if there is no support.

Just in case, I attach the error log

`[64/424] Linking CXX executable math_tests
FAILED: math_tests 
: && /usr/bin/c++ -Wall -Wextra -Wpedantic -Werror=return-type -Wno-unused -O3 -DNDEBUG  src/math/tests/CMakeFiles/math_tests.dir/math_tests_autogen/mocs_compilation.cpp.o src/math/tests/CMakeFiles/math_tests.dir/main.cpp.o src/math/tests/CMakeFiles/math_tests.dir/TestHessians.cpp.o src/math/tests/CMakeFiles/math_tests.dir/TestSqDistApproximant.cpp.o src/math/tests/CMakeFiles/math_tests.dir/TestMatrixCalc.cpp.o -o math_tests  src/math/libmath.a  /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.74.0  /usr/lib/x86_64-linux-gnu/libboost_prg_exec_monitor.so.1.74.0  src/foundation/libfoundation.a  /usr/lib/x86_64-linux-gnu/libQt5Xml.so.5.15.3  /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.15.3  /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.15.3 && :
/usr/bin/ld: src/math/tests/CMakeFiles/math_tests.dir/TestHessians.cpp.o: в функции «_GLOBAL__sub_I_TestHessians.cpp»:
TestHessians.cpp:(.text.startup+0x196): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::basic_cstring<char const>, boost::unit_test::basic_cstring<char const>, unsigned long, boost::unit_test::decorator::collector&)»
/usr/bin/ld: TestHessians.cpp:(.text.startup+0x21b): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestHessians.cpp:(.text.startup+0x2bb): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestHessians.cpp:(.text.startup+0x35b): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: src/math/tests/CMakeFiles/math_tests.dir/TestSqDistApproximant.cpp.o: в функции «_GLOBAL__sub_I_TestSqDistApproximant.cpp»:
TestSqDistApproximant.cpp:(.text.startup+0x196): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::basic_cstring<char const>, boost::unit_test::basic_cstring<char const>, unsigned long, boost::unit_test::decorator::collector&)»
/usr/bin/ld: TestSqDistApproximant.cpp:(.text.startup+0x21b): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestSqDistApproximant.cpp:(.text.startup+0x2bb): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestSqDistApproximant.cpp:(.text.startup+0x35b): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: src/math/tests/CMakeFiles/math_tests.dir/TestMatrixCalc.cpp.o: в функции «__static_initialization_and_destruction_0(int, int) [clone .constprop.0]»:
TestMatrixCalc.cpp:(.text.startup+0x18e): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::basic_cstring<char const>, boost::unit_test::basic_cstring<char const>, unsigned long, boost::unit_test::decorator::collector&)»
/usr/bin/ld: TestMatrixCalc.cpp:(.text.startup+0x21b): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestMatrixCalc.cpp:(.text.startup+0x2bb): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestMatrixCalc.cpp:(.text.startup+0x35b): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
/usr/bin/ld: TestMatrixCalc.cpp:(.text.startup+0x3fb): неопределённая ссылка на «boost::unit_test::ut_detail::auto_test_unit_registrar::auto_test_unit_registrar(boost::unit_test::test_case*, boost::unit_test::decorator::collector&, unsigned long)»
collect2: error: ld returned 1 exit status
[66/424] Building CXX object src/image.../imageproc.dir/ConnCompEraserExt.cpp.o
ninja: build stopped: subcommand failed.
`

tried commands
cmake -GNinja -S /home/user/scantailor-advanced-master_1.18.552 -B "/home/vlad/scantailor-advanced-master_1.18.552/Temp/build_scantailor-advanced_release"
cmake --build "/home/user/scantailor-advanced-master_1.18.552/Temp/build_scantailor-advanced_release"

and standard
cmake -G "Unix Makefiles" ..
cmake --build . --parallel ${nproc}

but there's a problem with Boost even though it's found
`
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found suitable version "1.74.0", minimum required is "1.60") found components: unit_test_framework prg_exec_monitor `

what should I do in Linux?

**Comment by trufanov-nok** [comment]
It does support build on Linux.  
Do you have a `/usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.74.0` file on drive?

**Reply by derrikF** [reply]
such a file exists

command "cmake .."
```
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found JPEG: /usr/lib/x86_64-linux-gnu/libjpeg.so (found version "80") 
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
-- Found PNG: /usr/lib/x86_64-linux-gnu/libpng.so (found version "1.6.37") 
-- Found TIFF: /usr/lib/x86_64-linux-gnu/libtiff.so (found version "4.3.0")  
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found suitable version "1.74.0", minimum required is "1.60") found components: unit_test_framework prg_exec_monitor 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE  
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so   
-- Found WrapOpenGL: TRUE  
-- Found XKB: /usr/lib/x86_64-linux-gnu/libxkbcommon.so (found suitable version "1.4.0", minimum required is "0.5.0") 
-- Found WrapVulkanHeaders: /usr/include  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vlad/scantailor-advanced-master_1.18.552/build
```
packages qt6-l10n-tools and qt6-tools-dev-tools are present... but not compiling

**Reply by trufanov-nok** [reply, marked as answer]
What `locate boost/test/unit_test.hpp` outputs?   
Perhaps you have multiple boost installations or corrupted installation. Try `sudo apt install --reinstall libboost-dev`

**Reply by derrikF** [reply]
reinstall didn't help 

```
$ locate boost/test/unit_test.hpp
/usr/include/boost/test/unit_test.hpp
/usr/local/include/boost/test/unit_test.hpp
```

**Reply by trufanov-nok** [reply]
I suspect that `/usr/include/boost/test/unit_test.hpp` and `/usr/local/include/boost/test/unit_test.hpp` are from different boost versions. You might have a locally built boost and TA-Lib build process gets a wrong header file.
What outputs: 
```
cat /usr/include/boost/version.hpp | grep "BOOST_LIB_VERSION"
cat /usr/local/include/boost/version.hpp | grep "BOOST_LIB_VERSION"
```
?

**Reply by derrikF** [reply]
$ cat /usr/include/boost/version.hpp | grep "BOOST_LIB_VERSION"
```
//  BOOST_LIB_VERSION must be defined to be the same as BOOST_VERSION
#define BOOST_LIB_VERSION "1_74"
```
$ cat /usr/local/include/boost/version.hpp | grep "BOOST_LIB_VERSION"
```
//  BOOST_LIB_VERSION must be defined to be the same as BOOST_VERSION
#define BOOST_LIB_VERSION "1_67"
```

**Reply by derrikF** [reply]
i removed boost
```
sudo apt-get -y --purge remove libboost-all-dev libboost-doc libboost-dev
sudo rm -r /usr/local/lib/libboost*
sudo rm -r /usr/local/include/boost
sudo rm -f /usr/lib/libboost_*
sudo rm -r /usr/include/boost
```
then installed two packages
libboost-test1.74-dev
libboost1.74-dev

and compilation is done

thank you nice man

**Comment by FriedrichFroebel** [comment]
I have used the following approach in the past, which still seems to work on Linux Mint 21, and is based upon the official building instructions residing at https://github.com/vigri/scantailor-libs-build#linux:

```bash
git clone https://github.com/vigri/scantailor-advanced scantailor-advanced_vigri/
cd scantailor-advanced_vigri
mkdir build
cd build
cmake ..
make
```

Your issues seem to be related to compiling the tests, which I never tried to compile or run.
